src/migrate.rs

```diff
@@ -0,0 +1,964 @@
+        cursor.seek(SeekFrom::Start(page_header_size(bits)))?;
+
+        Ok(MetaData {
+            mm_magic: cursor.read_u32::<LittleEndian>()?,
+            mm_version: cursor.read_u32::<LittleEndian>()?,
+            mm_address: cursor.read_uint::<LittleEndian>(bits.size())?,
+            mm_mapsize: cursor.read_uint::<LittleEndian>(bits.size())?,
+            mm_dbs: Databases {
+                free: Database::new(&mut cursor, bits)?,
+                main: Database::new(&mut cursor, bits)?,
+            },
+            mm_last_pg: cursor.read_uint::<LittleEndian>(bits.size())?,
+            mm_txnid: cursor.read_uint::<LittleEndian>(bits.size())?,
+        })
+    }
+
+    fn parse_leaf_nodes(cursor: &mut Cursor<&[u8]>, pb_lower: u16, bits: Bits) -> MigrateResult<Vec<LeafNode>> {
+        cursor.set_position(page_header_size(bits));
+        let num_keys = Self::num_keys(pb_lower, bits);
+        let mp_ptrs = Self::parse_mp_ptrs(cursor, num_keys)?;
+
+        let mut leaf_nodes = Vec::with_capacity(num_keys as usize);
+
+        for mp_ptr in mp_ptrs {
+            cursor.set_position(u64::from(mp_ptr));
+            leaf_nodes.push(Self::parse_leaf_node(cursor, bits)?);
+        }
+
+        Ok(leaf_nodes)
+    }
+
+    fn parse_leaf_node(cursor: &mut Cursor<&[u8]>, bits: Bits) -> MigrateResult<LeafNode> {
+        // The order of the mn_lo and mn_hi fields is endian-dependent and would
+        // be reversed in an LMDB environment created on a big-endian system.
+        let mn_lo = cursor.read_u16::<LittleEndian>()?;
+        let mn_hi = cursor.read_u16::<LittleEndian>()?;
+
+        let mn_flags = NodeFlags::from_bits(cursor.read_u16::<LittleEndian>()?).ok_or(MigrateError::InvalidNodeBits)?;
+        let mn_ksize = cursor.read_u16::<LittleEndian>()?;
+
+        let start = usize::try_from(cursor.position())?;
+        let end = usize::try_from(cursor.position() + u64::from(mn_ksize))?;
+        let key = cursor.get_ref()[start..end].to_vec();
+        cursor.set_position(end as u64);
+
+        let mv_size = Self::leaf_node_size(mn_lo, mn_hi);
+        if mn_flags.contains(NodeFlags::BIGDATA) {
+            let overflow_pgno = cursor.read_uint::<LittleEndian>(bits.size())?;
+
+            Ok(LeafNode::BigData {
+                mn_lo,
+                mn_hi,
+                mn_flags,
+                mn_ksize,
+                mv_size,
+                key,
+                overflow_pgno,
+            })
+        } else if mn_flags.contains(NodeFlags::SUBDATA) {
+            let start = usize::try_from(cursor.position())?;
+            let end = usize::try_from(cursor.position() + u64::from(mv_size))?;
+            let value = cursor.get_ref()[start..end].to_vec();
+            let mut cursor = std::io::Cursor::new(&value[..]);
+            let db = Database::new(&mut cursor, bits)?;
+            assert!(db.md_root != invalid_page_num(bits));
```

Comment:
Over the course of the parsing, shouldn't we just return with parsing error(s) instead of panicking? Ditto for all other assertions located in `get_meta_data()` and `migrate()`.

Reply:
Yes, you're right, we should return errors instead of panicking. I've replaced all the assertions with validation checks that return `MigrateError` variants on failure in 8a5867fbbb6482b9eb0662efbea62f5304b2a8b5.